### PR TITLE
Add support for Argo Rollout

### PR DESCRIPTION
**Pull request by kishoregv**
This pull request will fix https://github.com/hashicorp/vault-secrets-operator/issues/338

### CODE REVIEW

File: internal/helpers/rollout_restart.go [hunk 1]
```diff
@@ -135,6 +140,14 @@ func patchForRolloutRestart(ctx context.Context, obj ctrlclient.Object, client c
 		}
 		t.Spec.Template.ObjectMeta.Annotations[AnnotationRestartedAt] = time.Now().Format(time.RFC3339)
 		return client.Patch(ctx, t, patch)
+
+	case *rolloutsv1alpha1.Rollout:
+		patch := ctrlclient.MergeFrom(t.DeepCopy())
```
Comment (benashz) [inline internal/helpers/rollout_restart.go:145]: VSO will need to have the proper RBAC permissions to patch this new resource type. We will need to extend the RBAC role to include it. I think you can probably add:
```
//+kubebuilder:rbac:groups=argoproj.io,resources=rollouts,verbs=get;list;watch;patch
```

To each controller that handles secret sync. E.g: https://github.com/hashicorp/vault-secrets-operator/blob/648d116d0b8f2ef8a1a99a11a5c17fd1a0837f88/controllers/vaultdynamicsecret_controller.go#L66

We will need to update:
```
controllers/hcpvaultsecretsapp_controller.go
controllers/vaultdynamicsecret_controller.go
controllers/vaultpkisecret_controller.go
controllers/vaultstaticsecret_controller.go
```

File: internal/helpers/rollout_restart.go [hunk 2]
```diff
@@ -97,6 +98,10 @@ func RolloutRestart(ctx context.Context, namespace string, target v1beta1.Rollou
 		obj = &appsv1.StatefulSet{
 			ObjectMeta: objectMeta,
 		}
+	case "Rollout":
```
Comment (benashz) [inline internal/helpers/rollout_restart.go:101]:
```suggestion
	case "argo.Rollout":
```

**Review by benashz — COMMENTED** [review]
Thank you for your contribution to HashiCorp!

This PR looks pretty good. There a few thing to address before we can merge it. Please see my suggestions/comments for further details.

Thanks!

File: internal/helpers/rollout_restart.go (hunk 1 above)
Comment (kishoregv) [reply internal/helpers/rollout_restart.go:145]: Looks like this does not apply for `controllers/hcpvaultsecretsapp_controller.go` please correct me if I am missing something.

Comment (kishoregv) [reply internal/helpers/rollout_restart.go:145]: Updated also for `controllers/hcpvaultsecretsapp_controller.go`. Please let me know if you anymore feedback.

**Comment by kishoregv** [comment]
Hey @benashz, is it ok to merge this PR?

**Comment by Crazy-Hopper** [comment]
Hello!

Thank you very much for writing this PR!

I have a question/concern. Wouldn't patching the rollout object just trigger a new full-blown rollout instead of a plain pod restart ? Maybe we should somehow signal rollout controller instead ?

**Comment by Crazy-Hopper** [comment]
> Maybe we should somehow signal rollout controller instead ?

Oh! It's much simpler than that! Took a look at https://github.com/argoproj/argo-rollouts/blob/master/pkg/kubectl-argo-rollouts/cmd/restart/restart.go and it basically sets spec.restartAt field.

**Comment by kishoregv** [comment]
> Hello!
> 
> Thank you very much for writing this PR!
> 
> I have a question/concern. Wouldn't patching the rollout object just trigger a new full-blown rollout instead of a plain pod restart ? Maybe we should somehow signal rollout controller instead ?

Patch should do a restart. Do you see otherwise?

**Comment by kishoregv** [comment]
> Hello!
> 
> Thank you very much for writing this PR!
> 
> I have a question/concern. Wouldn't patching the rollout object just trigger a new full-blown rollout instead of a plain pod restart ? Maybe we should somehow signal rollout controller instead ?

This should do it. Do you see otherwise?

**Comment by kishoregv** [comment]
Hey @benashz, please let me know if there anything else you think that needs to be taken care before you can merge this PR

**Comment by Crazy-Hopper** [comment]
> Patch should do a restart. Do you see otherwise?

Yes, patch should do a restart. The question is restart of what exactly.

In my opinion, setting Spec.Template.ObjectMeta.Annotations[AnnotationRestartedAt] would trigger a new Rollout with all those strategy steps defined in the rollout being executed. (Just imagine that there are indefinite pauses in those steps...)

Instead, I'd recommend restarting just the current pods by setting spec.restartAt field of the Rollout to the current time.

**Comment by Crazy-Hopper** [comment]
I suggest it like this:

```patch
diff --git a/internal/helpers/rollout_restart.go b/internal/helpers/rollout_restart.go
index 907e8eb..12b9f65 100644
--- a/internal/helpers/rollout_restart.go
+++ b/internal/helpers/rollout_restart.go
@@ -143,10 +143,7 @@ func patchForRolloutRestart(ctx context.Context, obj ctrlclient.Object, client c

        case *rolloutsv1alpha1.Rollout:
                patch := ctrlclient.MergeFrom(t.DeepCopy())
-               if t.Spec.Template.ObjectMeta.Annotations == nil {
-                       t.Spec.Template.ObjectMeta.Annotations = make(map[string]string)
-               }
-               t.Spec.Template.ObjectMeta.Annotations[AnnotationRestartedAt] = time.Now().Format(time.RFC3339)
+               t.Spec.RestartAt = time.Now().Format(time.RFC3339)
                return client.Patch(ctx, t, patch)
        default:
                return fmt.Errorf("unsupported type %T for rollout-restart patching", t)
```

**Comment by Crazy-Hopper** [comment]
And by the way, it is possible for a Rollout object to lack template altogether. It might reference another object with the spec.workloadRef, thus trying to patch the template will not always work.

**Comment by kishoregv** [comment]
@benashz can you please help with getting this PR merged?

**Comment by thyton** [comment]
@kishoregv Thank you very much for your contribution! We're extending your work with integration test infrastructure and test cases for this feature in a separate PR. We'll link the PR here when it's ready.

**Comment by Crazy-Hopper** [comment]
OMG... Such a simple patch had to be moved from 0.6 (no estimate) to 0.7 (2x no estimate)???

**Comment by hsimon-hashicorp** [comment]
> OMG... Such a simple patch had to be moved from 0.6 (no estimate) to 0.7 (2x no estimate)???

Hi there! Since Vault is a pretty complex product with a lot of moving parts, even PRs that seem simple can have far-reaching consequences. As such, we do our best to make sure that PRs are implemented with, as Thy mentioned, as much supporting infrastucture (for example, integration tests and test cases) in an attempt to ensure there aren't unintended consequences when the PR is merged. This can take some extra time, as our engineers perform feature work in sprints in addition to engaging with the community on issues and PRs. 

Please take a look at our [community guidelines](https://www.hashicorp.com/community-guidelines) and keep in mind that the people who are developing these products are human beings and that we are not purposely delaying implementation. Thank you! :)

**Comment by thyton** [comment]
Closing this in favor of #702.

**Comment by Crazy-Hopper** [comment]
Lots of gratitudes to everyone involved for all the work done!

**Pull request closed without merging**